src/http.js

```diff
@@ -1,65 +1,73 @@
 import SDK_VERSION from './version';
-import STATUS from './status_codes';
-
-export function request(method, path, data, callback) {
-  const xhr = new XMLHttpRequest();
-
-  let url = `${API_HOST.getHost()}/${path}`;
-  let body = {};
-  if (method === 'GET') {
-    const qsArr = [];
-    for (let key in data) {
-      qsArr.push(`${key}=${data[key]}`);
-    }
-
-    if (qsArr.length > 0) {
-      const qs = encodeURIComponent(qsArr.join('&'));
-      url = `${url}?${qs}`;
-    }
-  } else {
-    body = data;
-  }
+import ERROR from './error_codes';
 
-  xhr.open(method, url, true);
+class Http {
+  static request(method, path, data) {
+    return new Promise((resolve, reject) => {
 
-  const publishableKey = Cookie.getCookie(Cookie.PUBLISHABLE_KEY);
-  if (!publishableKey) {
-    callback(STATUS.ERROR_PUBLISHABLE_KEY);
-    return;
-  }
+      const xhr = new XMLHttpRequest();
+
+      let url = `${API_HOST.getHost()}/${path}`;
+
+      // remove undefined values
+      let body = {};
+      Object.keys(data).forEach((key) => {
+        const value = data[key];
+        if (value !== undefined) {
+          body[key] = value;
+        }
+      });
+
+      // convert data to querystring for GET
+      if (method === 'GET') {
+        const params = Object.keys(body).map((key) => (
+          `${key}=${encodeURIComponent(body[key])}`
+        ));
 
-  xhr.setRequestHeader('Authorization', publishableKey);
-  xhr.setRequestHeader('Content-Type', 'application/json');
-  xhr.setRequestHeader('X-Radar-Device-Type', 'Web');
-  xhr.setRequestHeader('X-Radar-SDK-Version', SDK_VERSION);
-
-  xhr.onload = () => {
-    if (xhr.status == 200) {
-      try {
-        callback(STATUS.SUCCESS, JSON.parse(xhr.response));
-      } catch (e) {
-        callback(STATUS.ERROR_SERVER);
+        if (params.length > 0) {
+          const queryString = params.join('&');
+          url = `${url}?${queryString}`;
+        }
+
+        body = undefined; // dont send body for GET request
       }
-    } else if (xhr.status == 401) {
-      callback(STATUS.ERROR_UNAUTHORIZED);
-    } else if (xhr.status == 429) {
-      callback(STATUS.ERROR_RATE_LIMIT);
-    } else {
-      callback(STATUS.ERROR_SERVER);
-    }
-  }
 
-  xhr.onerror = function() {
-    callback(STATUS.ERROR_SERVER);
-  }
+      xhr.open(method, url, true);
 
-  xhr.timeout = function() {
-    callback(STATUS.ERROR_NETWORK);
-  }
+      const publishableKey = Cookie.getCookie(Cookie.PUBLISHABLE_KEY);
+      if (!publishableKey) {
+        reject(new Error(ERROR.PUBLISHABLE_KEY));
+        return;
+      }
 
-  xhr.send(JSON.stringify(body));
+      xhr.setRequestHeader('Authorization', publishableKey);
+      xhr.setRequestHeader('Content-Type', 'application/json');
+      xhr.setRequestHeader('X-Radar-Device-Type', 'Web');
+      xhr.setRequestHeader('X-Radar-SDK-Version', SDK_VERSION);
+
+      xhr.onload = () => {
+        try {
+          resolve(JSON.parse(xhr.response));
```

Comment:
doesn't look like we're checking `xhr.status` anymore? ex. the server could return a 429

Reply:
I don't think there's a need to check since it's just passing the error through to the caller now so it can inspect the api error (and get a better error message).

Reply:
true, we can probably prune the error codes to only include what JS can throw - doesn't look like we match against 401, 429, etc anymore